OPAM 2.0 does not understand symlinks

On my system directory `~/.opam` is a symlink to an actual directory on a different partition. This is due to lack of space on my `/home` partition.  OPAM 1.2.2 was perfectly fine with this setup, but OPAM 2.0 apparently does not understand symlinks and errors out.  Here's an example scenario. I have created an empty `~/.opam` directory (not a symlink), downloaded OPAM 2.0 with `sh <(curl -sL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh)` and installed 4.06.1 switch. All works fine so far. Now, I create a directory on a different partition, copy all of `~/.opam` contents to that directory, remove (now empty) `~/.opam`, and create a `~/.opam` symlink to the directory that now contains OPAM data.  Here's what happens when I try to install dune:

```
$ opam install dune
The following actions will be performed:
  ∗ install dune 1.4.0

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[dune.1.4.0] downloaded from cache at https://opam.ocaml.org/cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ERROR] The compilation of dune failed at "/home/killy/.opam/opam-init/hooks/sandbox.sh build ocaml bootstrap.ml".

#=== ERROR while compiling dune.1.4.0 =========================================#
# context     2.0.0 | linux/x86_64 | ocaml-base-compiler.4.06.1 | https://opam.ocaml.org#1fa4c078
# path        ~/.opam/4.06.1/.opam-switch/build/dune.1.4.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build ocaml bootstrap.ml
# exit-code   1
# env-file    ~/.opam/log/dune-31468-ad8886.env
# output-file ~/.opam/log/dune-31468-ad8886.out
### output ###
# Can't mkdir /home/killy/.opam/4.06.1: No such file or directory



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build dune 1.4.0
└─
╶─ No changes have been performed
```

I now revert everything to how it was originally, i.e. remove `~/.opam` symlink, create `~/.opam` directory, and copy back all of the contents that was originally there. Now everything works:

```
$ opam install dune
The following actions will be performed:
  ∗ install dune 1.4.0

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[dune.1.4.0] found in cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
∗ installed dune.1.4.0
Done.
```

```
$ opam config report
# opam config report
# opam-version      2.0.0
# self-upgrade      no
# system            arch=x86_64 os=linux os-distribution=debian os-version=9.5
# solver            builtin-mccs+glpk
# install-criteria  -removed,-count[version-lag,request],-count[version-lag,changed],-changed
# upgrade-criteria  -removed,-count[version-lag,solution],-new
# jobs              7
# repositories      1 (http) (default repo at 1fa4c078)
# pinned            0
# current-switch    4.06.1
```

--- a/src/core/opamFilename.ml
+++ b/src/core/opamFilename.ml
@@ -129,6 +129,10 @@ let (/) d1 s2 =
   let s1 = Dir.to_string d1 in
   raw_dir (Filename.concat s1 s2)
 
+let concat_and_resolve d1 s2 =
+  let s1 = Dir.to_string d1 in
+  Dir.of_string (Filename.concat s1 s2)
+
 type t = {
   dirname:  Dir.t;
   basename: Base.t;

--- a/src/core/opamFilename.mli
+++ b/src/core/opamFilename.mli
@@ -93,6 +93,9 @@ val with_tmp_dir: (Dir.t -> 'a) -> 'a
 (** Provide an automatically cleaned up temp directory to a job *)
 val with_tmp_dir_job: (Dir.t -> 'a OpamProcess.job) -> 'a OpamProcess.job
 
+(** Create a new Dir.t and resolve symlinks *)
+val concat_and_resolve: Dir.t -> string -> Dir.t
+
 include OpamStd.ABSTRACT
 
 (** Generic filename *)

--- a/src/state/opamStateConfig.ml
+++ b/src/state/opamStateConfig.ml
@@ -27,8 +27,8 @@ type t = {
 }
 
 let default = {
-  root_dir = OpamFilename.Op.(
-      OpamFilename.Dir.of_string (OpamStd.Sys.home ()) / ".opam"
+  root_dir = OpamFilename.(
+      concat_and_resolve (Dir.of_string (OpamStd.Sys.home ())) ".opam"
     );
   current_switch = None;
   switch_from = `Default;

